From dc084b409023b492edba1fc62efff7778dfcaa7b Mon Sep 17 00:00:00 2001
From: Chris Morgan <macromorgan@xxxxxxxxxxx>
Date: Thu, 16 Feb 2023 20:14:54 +0100
Subject: [PATCH] 9Tripod / Revise hardware tables

RTL8821CS is a WiFi + Bluetooth combo chip from Realtek that provides
WiFi A/B/G/N/AC over an SDIO interface and Bluetooth 4.2 over a UART
interface.

Note that the firmware this was tested with was firmware version
0xaa9a4e68 as reported by the driver. Attempts at using other
firmware were unsucessful.

Adapted for rockchip-5.10-rkr3.6 kernel.

Signed-off-by: Chris Morgan <macromorgan@xxxxxxxxxxx>
Signed-off-by: Steve Hay <me@stevenhay.com>
---
 drivers/bluetooth/btrtl.c  | 12 ++++++++++++
 drivers/bluetooth/hci_h5.c |  2 ++
 2 files changed, 14 insertions(+)

diff --git a/drivers/bluetooth/btrtl.c b/drivers/bluetooth/btrtl.c
index 3a9afc905f24..e3aeecaede0a 100644
--- a/drivers/bluetooth/btrtl.c
+++ b/drivers/bluetooth/btrtl.c
@@ -122,6 +122,18 @@ static const struct id_table ic_id_table[] = {
 	  .fw_name  = "rtl_bt/rtl8821a_fw.bin",
 	  .cfg_name = "rtl_bt/rtl8821a_config" },
 
+	/* 8821C with UART interface -- more specific match is first */
+	{ .match_flags = IC_MATCH_FL_LMPSUBV | IC_MATCH_FL_HCIREV |
+			 IC_MATCH_FL_HCIBUS,
+	  .lmp_subver = RTL_ROM_LMP_8821A,
+	  .hci_rev = 0xc,
+	  .hci_ver = 0x8,
+	  .hci_bus = HCI_UART,
+	  .config_needed = true,
+	  .has_rom_version = true,
+	  .fw_name  = "rtl_bt/rtl8821cs_fw.bin",
+	  .cfg_name = "rtl_bt/rtl8821cs_config" },
+
 	/* 8821C */
 	{ IC_INFO(RTL_ROM_LMP_8821A, 0xc),
 	  .config_needed = false,
diff --git a/drivers/bluetooth/hci_h5.c b/drivers/bluetooth/hci_h5.c
index 996729e78105..252c7a4b1a89 100644
--- a/drivers/bluetooth/hci_h5.c
+++ b/drivers/bluetooth/hci_h5.c
@@ -1022,6 +1022,8 @@ static const struct dev_pm_ops h5_serdev_pm_ops = {
 
 static const struct of_device_id rtl_bluetooth_of_match[] = {
 #ifdef CONFIG_BT_HCIUART_RTL
+	{ .compatible = "realtek,rtl8821cs-bt",
+	  .data = (const void *)&rtl_vnd },
 	{ .compatible = "realtek,rtl8822cs-bt",
 	  .data = (const void *)&rtl_vnd },
 	{ .compatible = "realtek,rtl8723bs-bt",
-- 
2.25.1

